src/v2/coins/abstractUtxoCoin.ts

```diff
@@ -833,46 +826,56 @@ class AbstractUtxoCoin extends BaseCoin {
       decompiledSigScript = currentInput.witness;
-      signatureScript = bitcoin.script.compile(decompiledSigScript);
-      if (isNativeSegwit) {
-        const lastWitness = _.last(transaction.ins[inputIndex].witness);
-        // const inputScriptHash = bitcoin.crypto.hash160(lastWitness);
-        const witnessScriptHash = bitcoin.crypto.sha256(lastWitness);
-        const prevOutScript = bitcoin.script.witnessScriptHash.output.encode(witnessScriptHash);
-        // we are faking a signature script for the verification
-        signatureScript = bitcoin.script.compile([prevOutScript]);
+      if (isNativeSegwitInput) {
+        inputClassification = bitcoin.script.classifyWitness(bitcoin.script.compile(decompiledSigScript), true);
+      } else {
+        inputClassification = bitcoin.script.classifyInput(currentInput.script, true);
       }
-    }
+    } else {
+      inputClassification = bitcoin.script.classifyInput(currentInput.script, true);
+      decompiledSigScript = bitcoin.script.decompile(currentInput.script);
 
-    const inputClassification = bitcoin.script.classifyInput(signatureScript, true);
-    if (inputClassification !== bitcoin.script.types.P2SH) {
-      if (inputClassification === bitcoin.script.types.P2PKH) {
-        const [signature, publicKey] = decompiledSigScript;
-        const publicKeys = [publicKey];
-        const signatures = [signature];
+    }
 
-        const pubScript = bitcoin.script.pubKeyHash.output.encode(bitcoin.crypto.hash160(publicKey));
-        return { signatures, publicKeys, isSegwitInput, inputClassification, pubScript: pubScript };
-      }
+    if (inputClassification === bitcoin.script.types.P2PKH) {
+      const [signature, publicKey] = decompiledSigScript;
+      const publicKeys = [publicKey];
+      const signatures = [signature];
+      const pubScript = bitcoin.script.pubKeyHash.output.encode(bitcoin.crypto.hash160(publicKey));
+
+      return { signatures, publicKeys, isSegwitInput, inputClassification, pubScript };
+    } else if (inputClassification === bitcoin.script.types.P2SH
+        || inputClassification === bitcoin.script.types.P2WSH) {
+      // Note the assumption here that if we have a p2sh or p2wsh input it will be multisig (appropriate because the
+      // BitGo platform only supports multisig within these types of inputs). Signatures are all but the last entry in
+      // the decompiledSigScript. The redeemScript/witnessScript (depending on which type of input this is) is the last
+      // entry in the decompiledSigScript (denoted here as the pubScript). The public keys are the second through
+      // antepenultimate entries in the decompiledPubScript. See below for a visual representation of the typical 2-of-3
+      // multisig setup:
+      //
+      // decompiledSigScript = 0 <sig1> <sig2> <pubScript>
+      // decompiledPubScript = 2 <pub1> <pub2> <pub3> 3 OP_CHECKMULTISIG
+      const signatures = decompiledSigScript.slice(0, -1);
```

Comment:
let's assert the correct signature and pubscript lengths here

Reply:
I'm not 100% sure we want to do that. Technically, prior to this someone could have been using this function to parse multisig p2sh inputs that had something other than a 2-of-3 configuration. Maybe SDK lord @tylerlevine could wiegh in here?

Reply:
at least we should check it is actually a valid multisig though

1. last item should be `OP_CHECKMULTISIG`
2. signature count should match `decompiledSigScript[0]`
3. pubScript count should match `decompileSigScript[-2]`

and then `verifySignature` could make sure it is `2-of-3` or I guess `3-of-3`

Reply:
`parseSignatureScript` should (ideally) be agnostic to the type of script it's parsing. @OttoAllmendinger's suggestion is correct though - the script should be validated and it's permissible to throw if the script is invalid.

Reply:
Done